from launch import LaunchDescription
from launch_ros.actions import Node
from ament_index_python.packages import get_package_share_directory
import os

def generate_launch_description():
    pkg = get_package_share_directory('uav_gnc')
    simulation_yaml = os.path.join(pkg, 'config', 'simulation.yaml')
    guidance_yaml = os.path.join(pkg, 'config', 'guidance.yaml')
    control_yaml = os.path.join(pkg, 'config', 'control.yaml')
    navigation_yaml = os.path.join(pkg, 'config', 'navigation.yaml')

    simulation = Node(
        package='uav_gnc',
        executable='simulation_node',   # CMake에서 만든 실행파일 이름
        name='simulation_node',
        output='screen',
        parameters=[simulation_yaml]
    )
    guidance = Node(
        package='uav_gnc',
        executable='guidance_node',
        name='guidance_node',
        output='screen',
        parameters=[guidance_yaml]
    )
    navigation = Node(
        package='uav_gnc',
        executable='navigation_node',
        name='navigation_node',
        output='screen',
        parameters=[navigation_yaml]
    )
    control = Node(
        package='uav_gnc',
        executable='control_node',
        name='control_node',
        output='screen',
        parameters=[control_yaml]
    )
    path_viz_sim = Node(
        package='uav_gnc',
        executable='path_viz_node',
        name='path_viz_sim',
        output='screen',
        parameters=[{
            'input_odom_topic': '/sim/odom',
            'output_path_topic': '/sim/path',
            'output_marker_topic': '/sim/marker',
            'publish_rate_hz': 10.0,
            'history_size': 5000
        }]
    )

    path_viz_nav = Node(
        package='uav_gnc',
        executable='path_viz_node',
        name='path_viz_nav',
        output='screen',
        parameters=[{
            'input_odom_topic': '/nav/odom',
            'output_path_topic': '/nav/path',
            'output_marker_topic': '/nav/marker',
            'publish_rate_hz': 10.0,
            'history_size': 5000
        }]
    )
    eval_sim = Node(
    package='uav_gnc',
    executable='tracking_eval_node',
    name='eval_sim',
    output='screen',
    parameters=[{
        'input_odom_topic': '/sim/odom',
        'csv_path': 'sim_tracking_eval.csv',
        'rate_hz': 20.0,
        'waypoints_x': [0.0, 5.0, 5.0, 0.0, 0.0],
        'waypoints_y': [0.0, 0.0, 5.0, 5.0, 0.0],
        'accept_radius': 0.5,
        }]
    )
    eval_nav = Node(
        package='uav_gnc',
        executable='tracking_eval_node',
        name='eval_nav',
        output='screen',
        parameters=[{
            'input_odom_topic': '/nav/odom',
            'csv_path': 'nav_tracking_eval.csv',
            'rate_hz': 20.0,
            'waypoints_x': [0.0, 5.0, 5.0, 0.0, 0.0],
            'waypoints_y': [0.0, 0.0, 5.0, 5.0, 0.0],
            'accept_radius': 0.5,
        }]
    )
    return LaunchDescription([
        simulation,
        guidance,
        navigation,
        control,
        path_viz_sim,
        path_viz_nav,
        eval_sim,
        eval_nav
    ])